src/rust/engine/fs/store/src/remote.rs

```diff
@@ -88,9 +89,70 @@ impl ByteStore {
     })
   }
 
-  pub async fn store_bytes(&self, bytes: Bytes) -> Result<Digest, String> {
-    let len = bytes.len();
+  pub(crate) fn chunk_size_bytes(&self) -> usize {
+    self.chunk_size_bytes
+  }
+
+  pub async fn store_buffered<WriteToBuffer, WriteResult>(
+    &self,
+    digest: Digest,
+    mut write_to_buffer: WriteToBuffer,
+  ) -> Result<(), String>
+  where
+    WriteToBuffer: FnMut(std::fs::File) -> WriteResult,
+    WriteResult: Future<Output = Result<(), String>>,
+  {
+    let write_buffer = tempfile::tempfile().map_err(|e| {
+      format!(
+        "Failed to create a temporary blob upload buffer for {digest:?}: {err}",
+        digest = digest,
+        err = e
+      )
+    })?;
+    let read_buffer = write_buffer.try_clone().map_err(|e| {
+      format!(
+        "Failed to create a read handle for the temporary upload buffer for {}",
+        e
+      )
+    })?;
+    write_to_buffer(write_buffer).await?;
+
+    // Unsafety: Mmap presents an immutable slice of bytes, but the underlying file that is mapped
+    // could be mutated by another process. We guard against this by creating an anonymous
+    // temporary file and ensuring it is written to and closed via the only other handle to it in
+    // the code just above.
+    let mmap = unsafe {
```

Comment:
I ended up using mmap not for performance reasons, but for fallibility reasons. The byte stream API expects a Stream<WriteRequest> which there is no way to produce when doing a standard io read on a file. Reads produce Results with a potential io::Error per each. The whole point is to not buffer up and so there is an impasse unless you're willing to expect Oks unchecked! With a memory mapping though, the failure happens up front at memory mapping time or not at all. Combined with the lazy semantics of memory maps and demand paging, this works out well. The madvise is for performance, although the result sof with / without I did not measure.

Reply:
You could use the [`async-stream` crate](https://crates.io/crates/async-stream) to easily create a stream from an async read loop. But this seems fine especially since the server will reject content that does not match the digest.

Reply:
Fwiw my impl was this:
```rust
#[derive(Debug)]
struct FileBuf {
  digest: Digest,
  file: tokio::fs::File,
  buf: BytesMut,
  chunk_size_bytes: usize,
  offset: usize,
}

impl FileBuf {
  fn new(digest: Digest, file: tokio::fs::File, chunk_size_bytes: usize) -> Self {
    Self {
      digest,
      file,
      buf: BytesMut::with_capacity(chunk_size_bytes),
      chunk_size_bytes,
      offset: 0,
    }
  }

  async fn advance(&mut self) -> Result<Option<Bytes>, std::io::Error> {
    if self.offset == self.digest.size_bytes {
      return Ok(None);
    }
    if self.buf.len() == self.chunk_size_bytes {
      return Ok(Some(self.buf.split().freeze()));
    }
    loop {
      let _read_bytes_len = self.file.read_buf(&mut self.buf).await?;
      if self.offset + self.buf.len() == self.digest.size_bytes
        || self.buf.len() >= self.chunk_size_bytes
      {
        self.offset += self.buf.len();
        let next_chunk = self.buf.split_off(self.chunk_size_bytes);
        let data = self.buf.clone().freeze();
        self.buf = next_chunk;
        return Ok(Some(data));
      }
    }
  }
}

pub async fn store_file<P: AsRef<Path>>(&self, digest: &Digest, path: P) -> Result<(), String> {
    let store_op = self.prepare_store_op(digest.to_owned());
    let resource_name = store_op.resource_name.clone();

    let len = digest.size_bytes;

    let file = tokio::fs::File::open(path)
      .map_err(|e| format!("{}", e))
      .await?;
    let mut file_buf = FileBuf::new(digest.to_owned(), file, self.chunk_size_bytes);
    let stream = async_stream::stream! {
        while let Some(data) = file_buf.advance().await.expect("TODO(John Sirois): XXX") {
            yield bazel_protos::gen::google::bytestream::WriteRequest {
              resource_name: resource_name.clone(),
              write_offset: (file_buf.offset - data.len()) as i64,
              finish_write: file_buf.offset == len,
              // TODO(tonic): Explore using the unreleased `Bytes` support in Prost from:
              // https://github.com/danburkert/prost/pull/341
              data
            };
        }
    };

    self.store_streaming(store_op, stream).await
  }
```

Note the TODO XXX. What to do about results when a Stream's values are not allowed to be fallible?

Reply:
This worked just fine, but ignoring the read errors  / panicing seemed unacceptable.

Reply:
To put a finer point on it: IIUC the generated rust client effectively forces one of:
1. Buffering all bytes in memory to form a complete Stream<WriteRequest>.
2. Using mmap for buffering.
3. Panicing if reading from any other fallible buffer like a standard file.

If that's wrong, I'd feel a bit better about getting rid of the unsafe.